pkg/storage/tsdb/upload.go

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package tsdb
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// UploadBlock is copy of block.Upload with following modifications:
+//
+// - If meta parameter is supplied (not nil), then local meta.json file is ignored, and uploaded meta.json file reflects meta parameter.
+//
+// - Meta struct is updated with gatherFileStats
+//
+// - external labels are not checked for
+func UploadBlock(ctx context.Context, logger log.Logger, bkt objstore.Bucket, blockDir string, meta *metadata.Meta) error {
+	df, err := os.Stat(blockDir)
+	if err != nil {
+		return err
+	}
+	if !df.IsDir() {
+		return errors.Errorf("%s is not a directory", blockDir)
+	}
+
+	// Verify dir.
+	id, err := ulid.Parse(df.Name())
+	if err != nil {
+		return errors.Wrap(err, "not a block dir")
+	}
+
+	if meta == nil {
+		meta, err = metadata.ReadFromDir(blockDir)
+		if err != nil {
+			// No meta or broken meta file.
+			return errors.Wrap(err, "read meta")
+		}
+	}
+
+	// Note that entry for meta.json file will be incorrect and will reflect local file,
+	// not updated Meta struct.
+	meta.Thanos.Files, err = gatherFileStats(blockDir)
+	if err != nil {
+		return errors.Wrap(err, "gather meta file stats")
+	}
+
+	metaEncoded := strings.Builder{}
+	if err := meta.Write(&metaEncoded); err != nil {
```

Comment:
```suggestion
	if err = meta.Write(&metaEncoded); err != nil {
```

Reply:
I prefer version with reduced scope of new `err`. Not sure I have a good reason though.